### PR TITLE
providers: add support for cloudstack config-drive

### PR DESCRIPTION
**Pull request by sgoeminn**
CloudStack config-drive support

Co-Authored-By: Sigert Goeminne <sigert.goeminne@nuagenetworks.net>

Upstream PR: https://github.com/apache/cloudstack/pull/2097
Design document: https://cwiki.apache.org/confluence/display/CLOUDSTACK/Using+ConfigDrive+for+Metadata%2C+Userdata+and+Password

### CODE REVIEW

File: internal/providers/cloudstack/cloudstack.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+func getPath(label string) (string, error) {
+	path := diskByLabelPath + label
+
+	if fileExists(path) {
+		return path, nil
+	}
+
+	out, err := exec.Command("blkid", "-l", "-t", "LABEL="+label, "-o", "device").Output()
```
Comment (crawford) [inline internal/providers/cloudstack/cloudstack.go]: Is there a reason you need to use `blkid` explicitly? udev should create all of the symlinks under `/dev/disk/by-label/`.

**Comment by larstobi** [comment]
@sgoeminn: nice work! I hope this gets merged 👍 

What do you think about @crawford's comment?

**Comment by sgoeminn** [comment]
@larstobi @crawford Thanks. Sorry for the wait, we are currently making some changes to the config drive implementation in CloudStack (based on feedback from the CloudStack conference in Miami). We will update the CoreOS PR's as soon as these changes are done.

File: internal/providers/cloudstack/cloudstack.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	for !labelExists(label) {
+		logger.Debug("config drive (%q) not found. Waiting...", label)
+		select {
+		case <-time.After(time.Second):
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+
+	path, err := getPath(label)
```
Comment (euank) [inline internal/providers/cloudstack/cloudstack.go]: this err isn't checked

Comment (sgoeminn) [inline internal/providers/cloudstack/cloudstack.go]: We added a check.

File: internal/providers/cloudstack/cloudstack.go (hunk 1 above)
Comment (sgoeminn) [inline internal/providers/cloudstack/cloudstack.go]: We added it as a backup. But normally it should not be needed indeed, so we removed it.

**Comment by crawford** [comment]
@sgoeminn Is there an easy way we can validate these changes? If I give you a test Container Linux image, would you be able to test it?

**Comment by larstobi** [comment]
Is this patch going to make v0.17.3?

**Comment by crawford** [comment]
@larstobi I'd like it to make it into the next release. We need a way to verify the changes though.

**Comment by sgoeminn** [comment]
@crawford Is manual testing sufficient? I already did manual testing with the first version of this PR. I can do this again with the updated version?

**Comment by larstobi** [comment]
I you can describe a sufficient testing procedure, I can also help with verification.

**Comment by sgoeminn** [comment]
@larstobi When you have a CloudStack setup based on this [PR 2097](https://github.com/apache/cloudstack/pull/2097), you just need to add a coreOS template based on this PR to CloudStack. After that, spin a VM in a network using the CoreOS template and provide UserData in ignition format as UserData through CloudStack. CloudStack should then create a ConfigDrive with label config-2 and execute the ignition configuration during boot. You can then ssh into the vm and verfiy if you are still able to mount the iso based on the label config-2 and check if ignition was executed with the config file yo specified (for example was a specific dummy file created, yes or no).

**Comment by larstobi** [comment]
@sgoeminn Oh, I only have XenServer to test with. :-( Which is why I'm basically only interested in the network service at coreos/bugs#1939. I can probably test that for you, if it would help?

**Comment by sgoeminn** [comment]
@larstobi Normally there will also be support for XenServer in the PR before it gets merged in into CloudStack. But currently, there is still some development work needed to get ConfigDrive working with XenServer.

**Comment by sgoeminn** [comment]
@crawford We successfully  tested ignition in combination with config drive on CloudStack.

File: internal/oem/oem.go [hunk 3]
```diff
@@ -138,6 +135,10 @@ func init() {
 			},
 		},
 	})
+	configs.Register(Config{name: "cloudstack",
```
Comment (crawford) [inline internal/oem/oem.go]: Can you move this back to its original location? They were in alphabetical order.

Can you also structure the code similar to the other definitions? The uniformity makes it easier to read.

File: internal/oem/oem.go [hunk 4]
```diff
@@ -138,6 +135,10 @@ func init() {
...
+		fetch:             cloudstack.FetchConfig,
+		defaultUserConfig: types.Config{Systemd: types.Systemd{Units: []types.Unit{userCloudInit("CloudStack", "cloudstack")}}},
```
Comment (crawford) [inline internal/oem/oem.go]: coreos-cloudinit does not support CloudStack. This default user config should be removed.

**Comment by sgoeminn** [comment]
@crawford I adapted the code.

**Comment by sgoeminn** [comment]
@crawford Done.

**Comment by crawford** [comment]
Thank you for your continued efforts. We are almost there.

**Pull request merged**